### DISCUSSION TITLE
unicast on a shared queue

### DISCUSSION THREAD

**Discussion by meravit** [rabbitmq/rabbitmq-server#6625] [Questions]
Hi,

Given a single queue that is shared by several consumers
1. Is it possible to address msg to specific consumer? 
   Note that I don't wish to use exchange as I don't want to have queue per consumer.
2. Let say msg contains the consumer Id. When consumer process msg, if the id doesn't match, requeue msg.
    Will msg eventually reach the right consumer? Is it guarantee RabbitMQ attempt sending msg to all consumers or we might have starvation?

**Comment by michaelklishin** [comment]
1. `rabbitmq_sharding` does not assume that you may need to address a specific consumer. It is technically possible since every shard is a queue whose name can be inspected, and if the queue name is known to the publisher, it can always use the default exchange.
2. It depends on what exchange is used with the sharded queue. It is a really inefficient way of going about this.

It sounds like you want `rabbitmq_consistent_has_exchange` where every queue is a "regular" one
and a group of messages with the same routing key would be routed to the same queue.

**Reply by meravit** [reply]
Hi Michael,
Thanks for you response.
Let me explain myself better.
I have one queue and several consumers that consume msgs from that queue. Every msg is consume by one of the consumer.
Now I wish to send msg that will be consumed by ALL consumers.   
I don't wanna use exchange and allocate queue per consumer (not good for me on high scale).
Is there a way to do broadcast on single queue?

My suggestions:
1.  send msg multiple times (according to consumers count), each with some consumer identifier. The consumers will check this identity - if no match the consumer, re-queue. 
or
2.  send msg multiple times (according to consumers count) , and assume that it will reach all due to round robin algorithm RabbitMQ use when choosing consumer to process msg. For higher insurance, I could send the msg several more times (I don't mind if consumer will get the msg twice)

What do you think? Can that work?

Merav

**Reply by mkuratczyk** [reply]
You can use a stream for the fan-out to all consumers messages. Each consumer would need to consume from two "queues" (one of them being a stream) but you can then send a single message and have all consumers read it from the stream, because stream consumption is non-destructive - the same message can be read as many times as you want. You can consume a stream via AMQP.
